src/session_pool/session_pool.js

```diff
@@ -0,0 +1,314 @@
+import log from 'apify-shared/log';
+import { checkParamOrThrow } from 'apify-client/build/utils';
+
+import { openKeyValueStore } from '../key_value_store';
+import { Session } from './session';
+import events from '../events';
+import { ACTOR_EVENT_NAMES_EX } from '../constants';
+
+
+/**
+ * Handles the sessions rotation, creation and persistence.
+ * Creates a pool of {@link Session} instances, that are randomly rotated.
+ * When some session is marked as blocked. It is removed and new one is created instead.
+ *
+ * Session pool is by default persisted in default {@link KeyValueStore}.
+ * If you want to have one pool for all runs you have to specify `persistStateKeyValueStoreId`.
+ *
+ * **Example usage:**
+ *
+ * ```javascript
+ * const sessionPool = new SessionPool({
+ *     maxPoolSize: 25,
+ *     maxSessionAgeSecs: 10,
+ *     maxSessionAgeSecs: 10,
+ *     maxSessionUsageCount: 150, // for example when you know that the site blocks after 150 requests.
+ *     persistStateKeyValueStoreId: 'my-key-value-store-for-sessions',
+ *     persistStateKey: 'my-session-pool',
+ * });
+ *
+ * // Now you have to initialize the `SessionPool`.
+ * // If you already have a persisted state in the selected `KeyValueState`.
+ * // The Session pool is recreated, otherwise it creates a new one.
+ * // It also attaches listener to `Apify.events` so it is persisted periodically and not after every change.
+ * await sessionPool.initialize();
+ *
+ * // Get random session from the pool
+ * const session1 = await sessionPool.retrieveSession();
+ * const session2 = await sessionPool.retrieveSession();
+ * const session3 = await sessionPool.retrieveSession();
+ *
+ * // Now you can mark the session either failed of successful
+ *
+ * // Fails session -> it increases error count (soft retire)
+ * session1.fail()
+ *
+ * // Marks as successful.
+ * session2.reclaim()
+ *
+ * // Retires session -> session is removed from the pool
+ * session3.retire()
+ *
+ * ```
+ */
+export class SessionPool extends EventEmitter {
+    /**
+     * Session pool configuration.
+     * @param options
+     * @param options.maxPoolSize {Number} - Maximum size of the pool.
+     * Indicates how many sessions are rotated.
+     * @param options.maxSessionAgeSecs {Number} - Number of seconds after which the session is considered as expired.
+     * @param options.maxSessionUsageCount {Number} - Maximum number of uses per session.
+     * It useful, when you know the site rate-limits, so you can retire the session before it gets blocked and let it cool down.
+     * @param options.persistStateKeyValueStoreId {String} - Name or Id of `KeyValueStore` where is the `SessionPool` state stored.
+     * @param options.persistStateKey {String} - Session pool persists it's state under this key in Key value store.
+     * @param options.createSessionFunction {function} - Custom function that should return `Session` instance.
+     */
+    constructor(options = {}) {
```

Comment:
Do we want people to use `new SessionPool`? If not, we should `@hideconstructor` (check if this goes to the class comment or the constructor comment)

Reply:
Yeah, we probably do not. I think there is no reason to don't use the openSessionPool shorthand.